🚀 4단계 - 수강신청(요구사항 변경)

baeksangha opened the pull request:
4단계  - 수강신청 (요구사항 변경) 코드입니다.
### 지난 step 피드백
- Period 클래스에 isBefore, isAfter를 사용해 validation 로직을 추가했습니다.
### DB 구현체 (JdbcRepository) 구조 변경 사항
- Course, Session, Student 클래스에 각각 List<Session>, List<Student>, List<Session> 필드가 존재했고 이를 repository에서 표현해주려다 보니 circular dependency가 발생하기도 하고 다대다 관계의 매핑이 고민되었습니다.
- 지난 step에서 피드백해주신대로 매핑 테이블을 만들고 SessionRepository에서 StudentRepository를 불러와 관련 정보를 얻었지만, 이번 step에서는 findById 로직에서 다른 도메인에 대한 정보를 넘겨주지 않도록 수정했습니다.
- 대신, 매핑 테이블을 서비스 레이어 단에서 조회하여 다른 도메인의 객체를 불러올 수 있게 하였고, 미래에 사용될 withSessions, withStudents 등의 메서드도 각 도메인에 구현해 두었습니다. (repository에서 다른 도메인의 정보를 불러오는 것은 바람직하지 않다고 생각했습니다.)
### 이번 step에서 변경된 사항
- 강의 진행 상태(준비중, 진행중, 종료)와 모집 상태(비모집중, 모집중)로 상태 값을 분리했습니다. 
  - 강의 진행 상태: Phase (READY("준비중"), IN_PROGRESS("진행중"), CLOSED("종료"))
  - 모집 상태: RecruitStatus (OPEN("모집중"), CLOSED("모집 마감"))
  - 두 상태를 합쳐 SessionStatus 객체로 표현했습니다.
- 하나의 강의에서 여러 개의 coverImage를 받을 수 있도록 수정했습니다.
  - 기존에 session 테이블에서 컬럼으로 가지고 있던 coverImage 관련 정보를 cover_image라는 새로운 테이블로 넣어주었습니다.
  - cover_image 테이블에는 필수적으로 session_id 정보를 가지고 있게 하였고, cover_image 테이블은 session 테이블에 종속적인 관계라고 판단하여 JdbcCoverImageDao를 만들어 JdbcSessionRepository에서 직접 이용하였습니다.
### 궁금한 점
마지막 요구사항을 명확히 이해하진 못했습니다 🥲
```
강사가 승인하지 않아도 수강 신청하는 모든 사람이 수강 가능하다.
우아한테크코스(무료), 우아한테크캠프 Pro(유료)와 같이 선발된 인원만 수강 가능해야 한다.
강사는 수강신청한 사람 중 선발된 인원에 대해서만 수강 승인이 가능해야 한다.
강사는 수강신청한 사람 중 선발되지 않은 사람은 수강을 취소할 수 있어야 한다.
```
Student 객체에 우테코, 우테코Pro로 선발된 인원인지 확인하는 PremiumPlan 필드를 추가했습니다.
Instructor 객체를 만들어 추후 어딘가에서 사용될 approve, decline 메서드를 만들었습니다.
요구 사항에서 보완해야할 점이 있다면 피드백 부탁드립니다.

지금까지 리뷰를 너무 잘 해주신 덕에 마지막 미션의 마지막단계까지 올 수 있었습니다. 4단계도 잘 부탁드립니다 🙇‍♂️

baeksangha committed:
feat: 강의 진행 상태와 모집 상태 분리 로직 추가

baeksangha committed:
refactor:
- Sessions를 id만 받는 형식에서 객체를 받는 형식으로 수정 (List<Long> -> List<Session>)
- Repository findById 시 객체값들은 반환하지 않도록 수정 (course_session과 session_student 테이블을 이용해 서비스 레이어의 어딘가에서 가져오도록 의도)

baeksangha committed:
feat: Session에서 하나 이상의 커버 이미지를 받도록 수정
- session 테이블에 있던 CoverImage 정보를 없애고 cover_image를 새 테이블로 추출
- JdbcCoverImageDao 생성 (session에 강결합 되어있다는 전제)
- 추후에 서비스 로직에서 findById를 이용할 때 CoverImage 정보를 넣어주는데 쓰일 withCoverImages 메서드 추가

baeksangha committed:
docs: README.md 업데이트

baeksangha committed:
feat: 학생이 우테코, 우테코pro에 선발되었는지 확인하는 필드 추가, instructor 클래스 추가

neojjc2 commented:
#tag @neojjc2

neojjc2 commented on README.md:
```diff
@@ -35,4 +35,21 @@
+### 프로그래밍 요구사항
+리팩터링할 때 컴파일 에러와 기존의 단위 테스트의 실패를 최소화하면서 점진적인 리팩터링이 가능하도록 한다.  
+DB 테이블에 데이터가 존재한다는 가정하에 리팩터링해야 한다.  
+즉, 기존에 쌓인 데이터를 제거하지 않은 상태로 리팩터링 해야 한다.
```
마지막 요구사항 정리 👍

neojjc2 commented:
> 4단계 - 수강신청 (요구사항 변경) 코드입니다.
> 
> ### 지난 step 피드백
> * Period 클래스에 isBefore, isAfter를 사용해 validation 로직을 추가했습니다.
> 
> ### DB 구현체 (JdbcRepository) 구조 변경 사항
> * Course, Session, Student 클래스에 각각 List, List, List 필드가 존재했고 이를 repository에서 표현해주려다 보니 circular dependency가 발생하기도 하고 다대다 관계의 매핑이 고민되었습니다.
> * 지난 step에서 피드백해주신대로 매핑 테이블을 만들고 SessionRepository에서 StudentRepository를 불러와 관련 정보를 얻었지만, 이번 step에서는 findById 로직에서 다른 도메인에 대한 정보를 넘겨주지 않도록 수정했습니다.
> * 대신, 매핑 테이블을 서비스 레이어 단에서 조회하여 다른 도메인의 객체를 불러올 수 있게 하였고, 미래에 사용될 withSessions, withStudents 등의 메서드도 각 도메인에 구현해 두었습니다. (repository에서 다른 도메인의 정보를 불러오는 것은 바람직하지 않다고 생각했습니다.)
> 

너무 좋습니다 👍 
네이밍도 좋아서 의도도 명확하게 잘 읽히는 것 같습니다 💯 


> ### 이번 step에서 변경된 사항
> * 강의 진행 상태(준비중, 진행중, 종료)와 모집 상태(비모집중, 모집중)로 상태 값을 분리했습니다.
>   
>   * 강의 진행 상태: Phase (READY("준비중"), IN_PROGRESS("진행중"), CLOSED("종료"))
>   * 모집 상태: RecruitStatus (OPEN("모집중"), CLOSED("모집 마감"))
>   * 두 상태를 합쳐 SessionStatus 객체로 표현했습니다.
> * 하나의 강의에서 여러 개의 coverImage를 받을 수 있도록 수정했습니다.
>   
>   * 기존에 session 테이블에서 컬럼으로 가지고 있던 coverImage 관련 정보를 cover_image라는 새로운 테이블로 넣어주었습니다.
>   * cover_image 테이블에는 필수적으로 session_id 정보를 가지고 있게 하였고, cover_image 테이블은 session 테이블에 종속적인 관계라고 판단하여 JdbcCoverImageDao를 만들어 JdbcSessionRepository에서 직접 이용하였습니다.
> 

네 이 부분도 좋습니다 👍 
조금 수정이 필요한 부분이 있는데 아래 코멘트로 보충하겠습니다 🙏 

> ### 궁금한 점
> 마지막 요구사항을 명확히 이해하진 못했습니다 🥲
> 
> ```
> 강사가 승인하지 않아도 수강 신청하는 모든 사람이 수강 가능하다.
> 우아한테크코스(무료), 우아한테크캠프 Pro(유료)와 같이 선발된 인원만 수강 가능해야 한다.
> 강사는 수강신청한 사람 중 선발된 인원에 대해서만 수강 승인이 가능해야 한다.
> 강사는 수강신청한 사람 중 선발되지 않은 사람은 수강을 취소할 수 있어야 한다.
> ```
> 
> Student 객체에 우테코, 우테코Pro로 선발된 인원인지 확인하는 PremiumPlan 필드를 추가했습니다. Instructor 객체를 만들어 추후 어딘가에서 사용될 approve, decline 메서드를 만들었습니다. 요구 사항에서 보완해야할 점이 있다면 피드백 부탁드립니다.
> 
> 지금까지 리뷰를 너무 잘 해주신 덕에 마지막 미션의 마지막단계까지 올 수 있었습니다. 4단계도 잘 부탁드립니다 🙇‍♂️

말씀하신 요구사항이 좀 모호한 면이 있네요 😄 
조금 더 부연 설명 드리자면, 

1. 일반 강의와 특별 강의(적당한 이름이 생각이 안나서 그냥 이렇게 부르겠습니다) 의 구분을 의미합니다. 
* 일반 강의 - 강사 승인 없이 누구나 수강 가능
* 특별 강의 - 우아한테크코스, 우아한테크캠프 Pro 등 선발 과정을 통과한 사람만 수강 가능

2. 그리고 특별 강의의 경우
* 강사는 "선발된 인원"에 대해서만 수강을 승인할 수 있음
* 강사는 "선발되지 않은 사람"의 수강 신청을 취소할 수 있음

이렇게 정리해볼 수 있을 것 같습니다 🙏 
그래서 지금 있는 `SessionType` 이외에 승인이 필요한 강의인지, 누구나 들을 수 있는 강의인지를 구분하는 값이 하나 필요할 것 같고
이미 만들어주신 `Instructor` 객체는 쓰이는 곳은 없지만 강의를 승인하거나, 취소하는 기능이 제공되면 될 것 같습니다. 
그리고 `PremiumPlan`는 잘 추가해주신 것 같은데 요구사항이 혼동스러우셔서 그런지 고민의 흔적이 남아있는 상태인데요 😄 
`Enum` 같은걸로 그냥 정의하고 수강이 가능한지, 불가능한지만 나타내는 정도로 정의되면 될 것 같습니다 🙇

neojjc2 commented:
> > 4단계 - 수강신청 (요구사항 변경) 코드입니다.
> > ### 지난 step 피드백
> > 
> > * Period 클래스에 isBefore, isAfter를 사용해 validation 로직을 추가했습니다.
> > 
> > ### DB 구현체 (JdbcRepository) 구조 변경 사항
> > 
> > * Course, Session, Student 클래스에 각각 List, List, List 필드가 존재했고 이를 repository에서 표현해주려다 보니 circular dependency가 발생하기도 하고 다대다 관계의 매핑이 고민되었습니다.
> > * 지난 step에서 피드백해주신대로 매핑 테이블을 만들고 SessionRepository에서 StudentRepository를 불러와 관련 정보를 얻었지만, 이번 step에서는 findById 로직에서 다른 도메인에 대한 정보를 넘겨주지 않도록 수정했습니다.
> > * 대신, 매핑 테이블을 서비스 레이어 단에서 조회하여 다른 도메인의 객체를 불러올 수 있게 하였고, 미래에 사용될 withSessions, withStudents 등의 메서드도 각 도메인에 구현해 두었습니다. (repository에서 다른 도메인의 정보를 불러오는 것은 바람직하지 않다고 생각했습니다.)
> > 
> > ### 이번 step에서 변경된 사항
> > 
> > * 강의 진행 상태(준비중, 진행중, 종료)와 모집 상태(비모집중, 모집중)로 상태 값을 분리했습니다.
> >   
> >   * 강의 진행 상태: Phase (READY("준비중"), IN_PROGRESS("진행중"), CLOSED("종료"))
> >   * 모집 상태: RecruitStatus (OPEN("모집중"), CLOSED("모집 마감"))
> >   * 두 상태를 합쳐 SessionStatus 객체로 표현했습니다.
> > * 하나의 강의에서 여러 개의 coverImage를 받을 수 있도록 수정했습니다.
> >   
> >   * 기존에 session 테이블에서 컬럼으로 가지고 있던 coverImage 관련 정보를 cover_image라는 새로운 테이블로 넣어주었습니다.
> >   * cover_image 테이블에는 필수적으로 session_id 정보를 가지고 있게 하였고, cover_image 테이블은 session 테이블에 종속적인 관계라고 판단하여 JdbcCoverImageDao를 만들어 JdbcSessionRepository에서 직접 이용하였습니다.
> > 
> > ### 궁금한 점
> > 마지막 요구사항을 명확히 이해하진 못했습니다 🥲
> > ```
> > 강사가 승인하지 않아도 수강 신청하는 모든 사람이 수강 가능하다.
> > 우아한테크코스(무료), 우아한테크캠프 Pro(유료)와 같이 선발된 인원만 수강 가능해야 한다.
> > 강사는 수강신청한 사람 중 선발된 인원에 대해서만 수강 승인이 가능해야 한다.
> > 강사는 수강신청한 사람 중 선발되지 않은 사람은 수강을 취소할 수 있어야 한다.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Student 객체에 우테코, 우테코Pro로 선발된 인원인지 확인하는 PremiumPlan 필드를 추가했습니다. Instructor 객체를 만들어 추후 어딘가에서 사용될 approve, decline 메서드를 만들었습니다. 요구 사항에서 보완해야할 점이 있다면 피드백 부탁드립니다.
> > 지금까지 리뷰를 너무 잘 해주신 덕에 마지막 미션의 마지막단계까지 올 수 있었습니다. 4단계도 잘 부탁드립니다 🙇‍♂️
> 
> 꼼꼼한 Description 너무 좋습니다 💯
> 
> 말씀하신 요구사항이 좀 모호한 면이 있네요 😄 조금 더 부연 설명 드리자면,
> 
> 1. 일반 강의와 특별 강의(적당한 이름이 생각이 안나서 그냥 이렇게 부르겠습니다) 의 구분을 의미합니다.
> 
> * 일반 강의 - 강사 승인 없이 누구나 수강 가능
> * 특별 강의 - 우아한테크코스, 우아한테크캠프 Pro 등 선발 과정을 통과한 사람만 수강 가능
> 
> 2. 그리고 특별 강의의 경우
> 
> * 강사는 "선발된 인원"에 대해서만 수강을 승인할 수 있음
> * 강사는 "선발되지 않은 사람"의 수강 신청을 취소할 수 있음
> 
> 이렇게 정리해볼 수 있을 것 같습니다 🙏 그래서 지금 있는 `SessionType` 이외에 승인이 필요한 강의인지, 누구나 들을 수 있는 강의인지를 구분하는 값이 하나 필요할 것 같고 이미 만들어주신 `Instructor` 객체는 쓰이는 곳은 없지만 강의를 승인하거나, 취소하는 기능이 제공되면 될 것 같습니다. 그리고 `PremiumPlan`는 잘 추가해주신 것 같은데 요구사항이 혼동스러우셔서 그런지 고민의 흔적이 남아있는 상태인데요 😄 `Enum` 같은걸로 그냥 정의하고 수강이 가능한지, 불가능한지만 나타내는 정도로 정의되면 될 것 같습니다 🙇

이건 개인적인 걱정인데, 막바지라고 너무 달리시는 것 같습니다 😅 거의 매일 새벽까지 병렬로 미션 
진행하시는 것 같은데, 건강을 해치실까 걱정이네요 😭 
거의 다 왔으니 조금 만 더 정리해주시면 될 것 같습니다 🙏 
화이팅입니다 💪

neojjc2 commented on src/main/java/nextstep/courses/domain/Instructor.java:
```diff
@@ -0,0 +1,47 @@
+            throw new IllegalArgumentException("유료 강의는 프리미엄 플랜이 필요합니다.");
+        }
+    }
+}
```
별도 코멘트에도 언급 드렸지만 도메인객체 생성 잘 해주셨습니다 😄 
`validation`도 좋구요 👍 
추가적으로 강의를 승인하거나 취소하는 기능만 제공되면 될 것 같습니다.

neojjc2 commented on src/main/java/nextstep/courses/domain/Period.java:
```diff
@@ -7,6 +7,12 @@ public class Period {
+        }
+        if (startDate.isBefore(LocalDate.now())) {
+            throw new IllegalArgumentException("시작일은 오늘 이후여야 합니다.");
+        }
```
검증 꼼꼼하게 개선해주셨네요 👍

neojjc2 commented on src/main/java/nextstep/courses/domain/PremiumPlan.java:
```diff
@@ -0,0 +1,19 @@
+    public boolean isWootecopro() {
+        return wootecopro;
+    }
+}
```
요 부분은 잘 정의해주셨는데 `Enum`정도로 정리하시고 강의를 수강할 수 있을지, 없을지는 
어딘가에 비즈니스로직으로 처리되는게 맞을 것 같습니다. 
그 결과를 `PASS`, `FAIL` 같은 식으로 `Student`에 가지고 있게 하면 될 것 같아요 😄 

좀 더 구체적으로 설명드리면 `Instructor`가 승인을 하면 `Student`에 `PASS`, 
취소를 하면 `FAIL` 같이 저장되면 
될 것 같습니다 🙇

neojjc2 commented on src/main/java/nextstep/courses/domain/RecruitStatus.java:
```diff
@@ -0,0 +1,11 @@
+    RecruitStatus(String description) {
+        this.description = description;
+    }
+}
```
💯 💯 💯 💯 💯

neojjc2 commented on src/main/java/nextstep/courses/domain/Session.java:
```diff
@@ -1,37 +1,36 @@
     protected final Long id;
     protected final Period period;
-    protected final CoverImage coverImage;
+    protected final CoverImages coverImages;
```
요구사항 구현 ✅

neojjc2 commented on src/main/resources/schema.sql:
```diff
@@ -54,19 +54,17 @@ create table student (
-    cover_height int not null,
-    status varchar(20) not null,
+    phase varchar(20) not null,
+    recruit_status varchar(20) not null,
```
컬럼 추가에 대한 `DDL`도 추가되면 좋을 것 같습니다 😄 

왜냐하면 4단계는 운영상태인점을 고려해서 개선해나가는게 요구사항이기 때문에
3단계 요구사항까지 배포된 상태로 추가 요구사항을 반영한다고 보시면 될 것 같습니다 🙇 

`phase varchar(20) not null;` 로 신규 컬럼이 추가되었는데 `not null` 이네요. 
(`recruit_status`도 동일) 
기존 테이블에 데이터들이 있다면 아마 이렇게 적용이 안될겁니다.
이러한 상황을 해결하기 위한 방법들이 있습니다:

#### 1. DEFAULT 값을 지정하여 컬럼 추가
```sql
ALTER TABLE session
ADD COLUMN phase varchar(20) NOT NULL DEFAULT 'READY';
````

#### 2. 또는 NULL을 허용하도록 컬럼을 추가한 후, 나중에 NOT NULL 제약조건을 추가:

```sql
-- 1단계: NULL을 허용하는 컬럼 추가 
ALTER TABLE session
ADD COLUMN phase varchar(20);

-- 2단계: 기존 데이터 업데이트
UPDATE session SET phase = 'READY' WHERE phase IS NULL;
-- 이 부분에서 조심해야 할 것이 기존 데이터를 READY으로 변경해도 되는지 체크가 필요합니다. 

-- 3단계: NOT NULL 제약조건 추가
ALTER TABLE session ALTER COLUMN phase SET NOT NULL;
```

그리고 위 내용과는 별도로 고민해주셔야 할 부분이 `SessionRepository`에서 
`SessionStatus`를 생성해주고 있는데요
```java
// 기존 데이터가 null이므로 이에 대한 고려가 필요합니다. 
SessionStatus status = new SessionStatus(
        Phase.valueOf(rs.getString("phase")),
        RecruitStatus.valueOf(rs.getString("recruit_status"))
);
```
`null` 일 경우 `NPE`가 발생할 수 있고 `null` 값이 아니라 임의의 값으로 처리한다 해도 
기존 로직에 영향없는지 파악이 필요합니다. 
```java
class PhaseTest {
    @DisplayName("null 값일 경우 NPE 발생")
    @Test
    void expectThrownNullPointerException() {
        assertThrows(NullPointerException.class, () -> Phase.valueOf(null));
    }
}
```
이 부분 보완이 좀 더 필요할 것 같습니다 🙇 

> 프로그래밍 요구사항
> * 리팩터링할 때 컴파일 에러와 기존의 단위 테스트의 실패를 최소화하면서 점진적인 리팩터링이 가능하도록 한다.
> * DB 테이블에 데이터가 존재한다는 가정하에 리팩터링해야 한다.
>  * 즉, 기존에 쌓인 데이터를 제거하지 않은 상태로 리팩터링 해야 한다.

neojjc2 reviewed (changes requested):
안녕하세요 상하님 🙇 

4단계 구현 하시느라 정말 고생 많으셨습니다 🙇 
리팩토링까지 같이 해주셔서
고생의 흔적이 많이 보이네요 😄 

마지막 단계이니만큼,
조금 정리되면 좋을만한 부분 의견 드렸습니다.
별도 코멘트로도 남겼는데요
한번 보시고 검토해주시면 감사하겠습니다 🙇

baeksangha committed:
refactor:
- Student에 수강 신청이 승인되었는지를 나타내는 필드(enrollStatus) 추가
- Session에서 해당 강의가 승인을 받아야 신청할 수 있는지, 아니면 바로 신청이 가능한지 나타내는 필드(participantType) 추가
- 관련 테스트 코드 추가

baeksangha committed:
refactor:
- schema.sql에 새롭게 추가되는 not null 컬럼(phase, recruit_status, enroll_status)값 default 추가

baeksangha commented:
준철님 안녕하세요. 마지막 리뷰까지 꼼꼼하게 봐주셔서 너무 감사합니다 🙇‍♂️😭🙇‍♂️ 새벽에 한건 제가 주말에 노느라 코딩을 안한 탓입니다..
새롭게 추가된 DDL에 대한 기존 데이터 마이그레이션 전략까지 고민하는 부분에서 정말 실무와 가깝다는 생각이 드네요.

리뷰 주신 부분 보고 아래 사항 수정했습니다.
- Student에 EnrollStatus라는 enum 클래스를 추가했습니다. 기본값은 NONE이고, 강사가 수강 승인을 했을 때 PASS, 거절당했을 때는 FAIL이 입력되도록 했습니다. 그리고 Session에는 해당 강의가 승인을 받아야 수강 등록을 할 수 있는 강의인지 여부를 나타내는 새로운 enum 클래스 ParticipantType을 추가하였습니다. 이에 맞추어 registerStudent 로직에 간단한 테스트 코드도 추가하였습니다.
- schema.sql에 새롭게 추가되는 컬럼에 대한 not null 정책을 고민해 보았습니다. 뭔가 코드 상에서 쿼리로 표현하기엔 한계가 있어 어떻게 하면 좋을지에 대한 고민을 해보았습니다.
  - stduent 클래스에 추가한 enroll_status같은 경우는, 2번 방법을 이용해 기존 로직에서 해당 student가 participant_type에 대한 신청 이력이 있을 경우 PASS로 처리한 후, default를 NONE으로 두면 좋을 것 같다고 생각했습니다.
  - session 클래스에 추가한 enum들도 전부 validation에도 사용되는 컬럼들이라 default를 선언하면 기존 데이터에 영향을 받을 것 같아 2번을 사용해야 할 것 같습니다. 다만 이게 쿼리 상으로 간단히 해결될 것 같진 않고 새로운 task를 만들어서 기존 row들에 대한 컬럼 값들을 채워준 후 default를 설정하면 좋을 것 같네요 (phase: 'READY', recruit_status: 'OPEN' 등)

`DB 테이블에 데이터가 존재한다는 가정하에 리팩터링해야 한다.`가 코드 상으로 표현하기 쉽지 않네요. 실무에서도 주로 새로운 task를 하나 만들어 기존 데이터를 마이그레이션했던 것 같습니다.

다시 리뷰 부탁드립니다. 항상 감사합니다 🙆‍♂️🙇‍♂️

neojjc2 commented on src/main/java/nextstep/courses/domain/EnrollStatus.java:
```diff
@@ -0,0 +1,5 @@
+
+public enum EnrollStatus {
+    PASS, FAIL, NONE
+}
```
💯 💯 💯 💯 💯

neojjc2 commented on src/main/java/nextstep/courses/domain/ParticipantType.java:
```diff
@@ -0,0 +1,5 @@
+
+public enum ParticipantType {
+    PRIVILEGED, NORMAL
+}
```
👍

neojjc2 commented on src/main/java/nextstep/courses/domain/PricingType.java:
```diff
@@ -0,0 +1,5 @@
+
+public enum PricingType {
+    FREE, PAID
+}
```
👍

neojjc2 commented on src/main/resources/schema.sql:
```diff
@@ -50,22 +50,22 @@ create table delete_history (
+    phase varchar(20) not null default 'READY',
+    recruit_status varchar(20) not null default 'OPEN',
+    pricing_type varchar(20),
+    participant_type varchar(20),
```
`not null` 까지 잘 신경써 주셨네요 👍

neojjc2 commented on src/main/java/nextstep/courses/domain/Session.java:
```diff
@@ -36,11 +36,20 @@ public Payment registerStudent(Student student) {
+            throw new IllegalArgumentException("승인 받은 수강생만 신청할 수 있는 강의입니다.");
+        }
+        validateBudgetOver(student);
+    }
```
좋습니다 💯

neojjc2 commented on src/main/java/nextstep/courses/domain/SessionType.java:
```diff
@@ -1,5 +1,19 @@
+
+    public ParticipantType getParticipantType() {
+        return participantType;
+    }
```
기존 `Enum`을 조합한 `SessionType`도 너무 좋네요 👍

neojjc2 commented:
> 준철님 안녕하세요. 마지막 리뷰까지 꼼꼼하게 봐주셔서 너무 감사합니다 🙇‍♂️😭🙇‍♂️ 새벽에 한건 제가 주말에 노느라 코딩을 안한 탓입니다.. 새롭게 추가된 DDL에 대한 기존 데이터 마이그레이션 전략까지 고민하는 부분에서 정말 실무와 가깝다는 생각이 드네요.
> 
> 리뷰 주신 부분 보고 아래 사항 수정했습니다.
> 
> * Student에 EnrollStatus라는 enum 클래스를 추가했습니다. 기본값은 NONE이고, 강사가 수강 승인을 했을 때 PASS, 거절당했을 때는 FAIL이 입력되도록 했습니다. 그리고 Session에는 해당 강의가 승인을 받아야 수강 등록을 할 수 있는 강의인지 여부를 나타내는 새로운 enum 클래스 ParticipantType을 추가하였습니다. 이에 맞추어 registerStudent 로직에 간단한 테스트 코드도 추가하였습니다.
> * schema.sql에 새롭게 추가되는 컬럼에 대한 not null 정책을 고민해 보았습니다. 뭔가 코드 상에서 쿼리로 표현하기엔 한계가 있어 어떻게 하면 좋을지에 대한 고민을 해보았습니다.
>   
>   * stduent 클래스에 추가한 enroll_status같은 경우는, 2번 방법을 이용해 기존 로직에서 해당 student가 participant_type에 대한 신청 이력이 있을 경우 PASS로 처리한 후, default를 NONE으로 두면 좋을 것 같다고 생각했습니다.
>   * session 클래스에 추가한 enum들도 전부 validation에도 사용되는 컬럼들이라 default를 선언하면 기존 데이터에 영향을 받을 것 같아 2번을 사용해야 할 것 같습니다. 다만 이게 쿼리 상으로 간단히 해결될 것 같진 않고 새로운 task를 만들어서 기존 row들에 대한 컬럼 값들을 채워준 후 default를 설정하면 좋을 것 같네요 (phase: 'READY', recruit_status: 'OPEN' 등)
> 

다소 두서 없는 리뷰인데, 이렇게 고민해주시고 찰떡같이 이해해주셔서 너무 좋네요 😄 
개선 너무 잘 해주셨습니다 💯 

> `DB 테이블에 데이터가 존재한다는 가정하에 리팩터링해야 한다.`가 코드 상으로 표현하기 쉽지 않네요. 실무에서도 주로 새로운 task를 하나 만들어 기존 데이터를 마이그레이션했던 것 같습니다.
> 

네 맞습니다. 서비스를 하고 있는 상태라면 항상 기존 데이터를 생각해야 하는데요, 데이터 마이그레이션도 쉽지 않은게 
데이터 양이 많다면 (몇 천만건 이상) 부담스러운게 사실입니다 😅 그래서 경우에 따라 코드에서 처리하기도 하는데요
이게 또 시간이 지나면 레거시 폭탄이 되서 아무리 테이블을 봐도 없는 데이터인데 어디에서 가져와서 사용되는거지?? 하고
한참을 찾게 됩니다 😭 물론 어쩔수 없는 상황에서 최선의 선택을 했을 거라 생각합니다. 그렇기 때문에 다양한 케이스에 대해서
종합적으로 문제해결을 할 수 있는 방법을 늘 고민해야 하는게 서버엔지니어의 역량 중에 하나라고 생각합니다. 

마지막이라 그런지 약간 꼰대(?) 같았네요 😄 개인적인 의견이니 참고 정도만 해주시면 될 것 같습니다 🙏 
너무 수고 많으셨습니다 🙇 

> 다시 리뷰 부탁드립니다. 항상 감사합니다 🙆‍♂️🙇‍♂️

neojjc2 reviewed (approved):
안녕하세요 상하님 🙇 

개선사항까지 깔끔하게 정리해주셨네요 👍 
소소한 의견 코멘트로 남겨드렸는데요
한번 읽어봐만 주시면 될 것 같습니다 🙇 

마지막까지 너무 괴롭혀드린게 아닌가 걱정이었는데
깊이 고민해주시고 잘 따라와 주셔서
미션 잘 마무리 할 수 있었던 것 같습니다. 

그동안 미션하시느라 정말 고생 많으셨습니다. 
앞으로 가실 개발 커리어도 더욱 발전하시기를 소망합니다.

이제 푹 주무실 수 있겠네요 😅 
`merge` 하겠습니다. 

LGTM 👍
